src/web100srv.c

```diff
@@ -1466,74 +1352,120 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
   ctl.socket = ctlsockfd;
   // this is the child process, it handles the connection with the client and
   // runs the actual tests.
-  log_println(4, "Child thinks pipe() returned fd0=%d, fd1=%d for pid=%d",
-              chld_pipe[0], chld_pipe[1], chld_pid);
-  close(listenfd);
-  close(chld_pipe[1]);
+  // First, start the watchdog timer
+  alarm(300);
+
+  // Set up the connection with the client (with optional SSL and websockets)
+  if (options.tls) {
+    ctl.ssl = SSL_new(ssl_context);
+    if (ctl.ssl == NULL) {
+      log_println(4, "SSL_new failed");
+      exit(-1);
+    }
+    if (SSL_set_fd(ctl.ssl, ctl.socket) == 0) {
+      log_println(4, "SSL_set_fd failed");
+      exit(-1);
+    }
+    if ((ssl_err = SSL_accept(ctl.ssl)) != 1) {
+      if (ssl_err == 0) {
+        ssl_err = SSL_get_error(ctl.ssl, ssl_err);
+      }
+      ERR_error_string_n(ssl_err, ssl_error, sizeof(ssl_error));
+      log_println(4, "SSL_accept failed (%s)", ssl_error);
+      exit(-1);
+    }
+  }
+
+  t_opts = initialize_tests(&ctl, &testopt, test_suite, sizeof(test_suite));
+  if (t_opts < 1) {  // some error in initialization routines
+    log_println(3, "Invalid test suite received, terminate child");
+    exit(-1);
+  }
+
 #if USE_WEB100
   if ((agent = web100_attach(WEB100_AGENT_TYPE_LOCAL, NULL)) == NULL) {
     web100_perror("web100_attach");
-    return 1;
+    exit(1);
   }
 #elif USE_WEB10G
   if (estats_nl_client_init(&agent) != NULL) {
-    log_println(0,
-                  "Error: estats_client_init failed."
-                  "Unable to use web10g.");
-    return 1;
+    log_println(0, "Error: estats_client_init failed. Unable to use web10g.");
+    exit(1);
   }
 #endif
-  // This is the child process from the above fork().  The parent
-  //  is in control, and will send this child a signal when it gets
-  //  to the head of the testing queue.  Post a read() and simply
-  //  wait for the parent to let us know it's time to move on.
-  //  Rich Carlson 3/11/04
   for (;;) {
-    memset(buff, 0, sizeof(buff));
-
     // the read() could return if an interrupt was caught.  This
     // condition should be checked for and the read() restarted if
     // necessary
-    // RAC 3/18/10
-    //
-    retcode = read(chld_pipe[0], buff, 32);
-    log_println(6, "Child %d received '%s' from parent", getpid(),
-                buff);
-    if ((retcode == -1) && (errno == EINTR))
-      continue;
-    if (strncmp(buff, "go", 2) == 0) {
-      log_println(6, "Got 'go' signal from parent, ready to start "
-                  "testing %d", getpid());
+    retcode = read(parent_pipe, &parent_message, sizeof(int));
+    log_println(6, "Child %d got '%d' from parent", getpid(), parent_message);
+    if ((retcode == -1) && (errno == EINTR)) continue;
+    if (parent_message == SRV_QUEUE_TEST_STARTS_NOW) {
+      send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_TEST_STARTS_NOW_STR,
+                            strlen(SRV_QUEUE_TEST_STARTS_NOW_STR),
```

Comment:
Oh nevermind, I grep'ed wrong and though these weren't defined in my view of the code.